src/infra/core/src/repos/metadata_chain_validators.rs

```diff
@@ -0,0 +1,338 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use chrono::{DateTime, Utc};
+use kamu_core::{
+    AppendError,
+    AppendValidationError,
+    BlockNotFoundError,
+    HashedMetadataBlockRef,
+    OffsetsNotSequentialError,
+    SequenceIntegrityError,
+};
+use opendatafabric::{
+    IntoDataStreamBlock,
+    IntoDataStreamEvent,
+    MetadataBlock,
+    MetadataBlockDataStreamRef,
+    MetadataEvent,
+    Multihash,
+};
+
+use crate::{invalid_event, Decision, MetadataBlockTypeFlags, MetadataChainVisitor};
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct ValidateSeedBlockOrderVisitor<'a> {
+    initial_block: &'a MetadataBlock,
+}
+
+impl<'a> ValidateSeedBlockOrderVisitor<'a> {
+    pub fn new((_, block): HashedMetadataBlockRef<'a>) -> Self {
+        Self {
+            initial_block: block,
```

Comment:
```rust
...
pub fn new(block: &MetadataBlock) -> Result<Self, AppendError> {
...
```

That's exactly the solution I was thinking about this weekend -- it would really simplify it. 

Will be done

Reply:
Done

The only difference is that the ctor also returns a `Decision`